main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala

```diff
@@ -1873,26 +1873,47 @@ object ParsedAst {
     */
   case class Annotation(sp1: SourcePosition, ident: Name.Ident, sp2: SourcePosition)
 
+
   /**
-    * An enum representing the handler of a `try` expression.
+    * A common super-type for lists of handlers.
     */
-  sealed trait CatchOrHandler
+  sealed trait HandlerList
+
+  object HandlerList {
+
+    /**
+      * A wrapper for a list of [[TryHandler.Catch]]es. This ensures that only catch-handlers are in the list.
+      *
+      * @param handlers The list of [[TryHandler.Catch]]
+      */
+    case class CatchHandlerList(handlers: Seq[TryHandler.Catch]) extends HandlerList
+
+    /**
+      * A wrapper for a list of [[TryHandler.WithHandler]]es. This ensures that only with-handlers are in the list.
+      *
+      * @param handlers The list of [[TryHandler.WithHandler]]
+      */
+    case class WithHandlerList(handlers: Seq[TryHandler.WithHandler]) extends HandlerList
+
+  }
+
+  object TryHandler {
 
-  object CatchOrHandler {
     /**
       * A `catch` block for handling Java exceptions.
       *
       * @param rules the catch rules.
       */
-    case class Catch(rules: Seq[ParsedAst.CatchRule]) extends CatchOrHandler
+    case class Catch(rules: Seq[ParsedAst.CatchRule])
 
     /**
       * A `with` block for handling Flix effects.
       *
       * @param eff   the effect to be handled.
       * @param rules the handler rules.
       */
-    case class Handler(eff: Name.QName, rules: Option[Seq[ParsedAst.HandlerRule]]) extends CatchOrHandler
+    case class WithHandler(eff: Name.QName, rules: Option[Seq[ParsedAst.HandlerRule]])
```

Comment:
Why are rules here optional?

Reply:
Not sure, it was like this before. Maybe because you can have an empty handler?

Reply:
IMO don't complicate the parser to make types nice in ParsedAst.

Reply:
But why is it optional in the parser?

Reply:
Because you can do `with E` (no braces) when `E` has no ops

Reply:
I guess it's a design decision. Is an empty effect like an empty enum?

Reply:
I would drop it. We can reintroduce if the need arise.

With empty enums there was actually a use case.

Reply:
Apparently we allowed this
```scala
// Test.Exp.Effect.flix
    // @test
    pub def handleEff03(): Bool = {
        try true with Empty
    }
```

Reply:
Yeah, I don't think we should do that :)